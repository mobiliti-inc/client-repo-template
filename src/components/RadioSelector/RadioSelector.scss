@import '~variables';

.data-selector {
  display: flex;
  cursor: pointer;
  justify-content: space-between;
  padding: 0 10%;
  @media (max-width: $half-mobile-width) {
    padding: 0 5%;
	}
  .key {
    display: flex;
    justify-content: center;
    align-items: center;
    .radio-button {
      display: flex;
      border: 1px solid #B9B9B9;
      width: 25px;
      height: 25px;
      border-radius: 50%;
    }
    .radio-button-selected {
      width: 15px;
      height: 15px;
      border-radius: 50%;
      margin: auto;
      background: $blue;
      transition: background 0.25s linear;
      -webkit-transition: background 0.25s linear;
    }

    .radio-button.selected {
      border: 1px solid $blue;
    }

    .radio-button.disabled {
      border: 1px solid #8F8F8F;

    }

    .key-text {
      padding-left: 17px;
      font-family: $hanken-sans;
      color: $darker-blue;
      font-size: 16px;
      opacity: 0.8;
    }

    .key-text.disabled {
      color: #8F8F8F;

    }
  }
}

.disabled {
  cursor: auto;
}

.data-selector.selected {
  background: #ECF4F6;
  margin: 0;
}

.value {
  font-weight: bold;
  font-size: 16px;
  font-family: $hanken-sans;
  color: $darker-blue;
}

.value.selected {
  padding-right: 0;
}

.value.disabled {
  color: #8F8F8F;
}

.hide {
  display: none;
}


.radio-icon {
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  width: 33px;
  height: 21px;
  left: 722px;
  top: 831px;
  margin-left: 20px;
}

.card-number {
	font-family: $hanken-sans;
  font-style: normal;
  font-weight: normal;
  line-height: normal;
  color: $darker-blue;
}

.all-cards {
  padding-left: 20px;
  font-family: $hanken-sans;
  color: $darker-blue;
}

.optional-text {
  margin-left: 20px;
  font-family: $hanken-sans;
  color: $darker-blue;
  font-size: 16px;
  opacity: 0.8;
}