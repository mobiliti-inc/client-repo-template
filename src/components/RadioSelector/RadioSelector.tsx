/* eslint-disable react/no-did-update-set-state */
import React from 'react';
import classNames from 'classnames';

import styles from './RadioSelector.scss';

class RadioSelector extends React.PureComponent {
	static propTypes = {
		onSelectOption: PropTypes.func.isRequired,
		options: PropTypes.oneOfType([
			PropTypes.arrayOf(PropTypes.shape({
				key: PropTypes.oneOfType([
					PropTypes.string,
					PropTypes.number
				]),
				value: PropTypes.oneOfType([
					PropTypes.string,
					PropTypes.number
				])
			})),
			PropTypes.arrayOf(PropTypes.shape({
				value: PropTypes.oneOfType([
					PropTypes.string,
					PropTypes.number
				])
			}))
		]).isRequired,
		disabled: PropTypes.bool,
		hideValue: PropTypes.bool,
		keyIsVisible: PropTypes.bool,
		selectedItem: PropTypes.string,
	}

	static defaultProps = {
		disabled: false,
		hideValue: false,
		keyIsVisible: true,
		selectedItem: '',
	}
	constructor(props) {
		super(props);
		this.state = {
			selectedIndex: null,
			data: {},
			options: props.options
		};
	}

	componentWillMount() {
		const { selectedItem } = this.props;
		this.setState({
			data: { key: selectedItem }
		});
	}

	componentDidUpdate(prevProps) {
		const { options, selectedItem } = this.props;
		if (options.length !== prevProps.options.length) {
			this.setState({
				options,
			});
		}

		if (prevProps.selectedItem !== selectedItem) {
			this.setState({
				data: { key: selectedItem }
			});
		}
	}

	onSelect = (event, data, index) => {
		const { selectedIndex } = this.state;
		const { options, disabled } = this.props;
		const optionDisabled = data.disabled;
		const radioDisabled = disabled;
		if (optionDisabled || radioDisabled) {
			event.stopPropagation();
		} else {
			if (selectedIndex) { Object.assign(options[selectedIndex], { current: false }); }
			Object.assign(options[index], { current: true });
			Object.assign({}, options);
			Object.assign(data, { current: true });
			this.setState({
				selectedIndex: index,
				data,
				options
			});
			this.props.onSelectOption(data, index);
		}
	}

	checkSeleceted = (selectedData) => {
		const { data } = this.state;
		if (selectedData.key) {
			return selectedData.current && data.key === selectedData.key;
		}
	}

	renderOptions = (options) => options.map((data, index) => {
		const { key, value, imageUrl, optionalText } = data;
		const { disabled, hideValue, keyIsVisible } = this.props;
		const optionDisabled = data.disabled;
		const radioDisabled = disabled;
		const radioButtonClass = classNames('radio-button', { selected: this.checkSeleceted(data) });
		const dataSelectorClass = classNames('data-selector', { selected: this.checkSeleceted(data), disabled: optionDisabled || radioDisabled });
		const valueClass = classNames('value', { selected: this.checkSeleceted(data), disabled: optionDisabled || radioDisabled, hide: hideValue });
		const keyTextClass = classNames('key-text', { disabled: optionDisabled || radioDisabled });
		return (
			<div
				role="presentation"
				styleName={dataSelectorClass}
				onClick={(event) => this.onSelect(event, data, index)}
				key={key}
			>
				<div styleName="key">
					<div styleName={radioButtonClass}>
						<div styleName={this.checkSeleceted(data) ? 'radio-button-selected' : 'hide'} />
					</div>
					{keyIsVisible && <p styleName={keyTextClass}>{key}</p> }
					{imageUrl && <img styleName="radio-icon" src={imageUrl} alt="radio icon" />}
					<div styleName="optional-text">
						{ optionalText }
					</div>
				</div>
				{value && <p styleName={valueClass}>{value}</p>}
			</div>
		);
	})

	render() {
		const { options } = this.state;
		return (
			<React.Fragment>
				{this.renderOptions(options)}
			</React.Fragment>
		);
	}
}

export default CSSModules(RadioSelector, styles, { allowMultiple: true });
