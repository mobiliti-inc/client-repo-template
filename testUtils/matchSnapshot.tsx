const { shallowToJson } = require('enzyme-to-json');

export const matchSnapshot = (subject: any) => {
	expect(shallowToJson(subject)).toMatchSnapshot();
};
