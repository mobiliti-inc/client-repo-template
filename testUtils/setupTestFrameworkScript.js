const React = require("react");
// const CSSModules = require("react-css-modules");
const Enzyme = require("enzyme");
const Adapter = require("enzyme-adapter-react-16");

const { shallow, render, mount } = Enzyme;

Enzyme.configure({ adapter: new Adapter() });
jest.setTimeout(30000);

global.React = React;
// global.CSSModules = CSSModules;
global.shallow = shallow;
global.render = render;
global.mount = mount;
