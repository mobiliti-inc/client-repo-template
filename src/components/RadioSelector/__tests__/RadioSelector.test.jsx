import RadioSelector from '../RadioSelector';
import { matchSnapshot } from '../../../../testUtils/matchSnapshot';

describe('RadioSelector', () => {
	const options = [
		{
			miles: 2000,
			status: 1,
			price: 1633,
			mobiliti_fee: null,
			rights: 0,
			key: '2000 Miles',
			value: '$1633',
			current: false
		},
		{
			miles: 2000,
			status: 1,
			price: 1633,
			mobiliti_fee: null,
			rights: 0,
			key: '200 Miles',
			value: '$633',
			current: false
		}
	];
	it('renders', () => {
		const subject = shallow(<RadioSelector
			options={options}
			onSelectOption={jest.fn()}
			selectedItem=" "
		/>);
		matchSnapshot(subject);
	});

	it('sets the correct value when onSelect method is called', () => {
		const subject = shallow(
			<RadioSelector
				options={options}
				onSelectOption={jest.fn()}
				selectedItem=" "
			/>
		);

		const data = {
			miles: 2000,
			status: 1,
			price: 1633,
			mobiliti_fee: null,
			rights: 0,
			key: '200 Miles',
			value: '$633',
			current: false
		};

		const event = {
			stopPropagation: jest.fn()
		};

		expect(subject.instance().state.selectedIndex).toEqual(null);
		subject.instance().onSelect(event, data, 1);
		expect(subject.instance().state.selectedIndex).toEqual(1);
		expect(subject.instance().state.data).toEqual(data);
	});
});
