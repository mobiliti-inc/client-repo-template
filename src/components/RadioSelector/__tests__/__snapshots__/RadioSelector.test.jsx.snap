// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioSelector renders 1`] = `
<React.Fragment>
  <div
    className="data-selector"
    key="2000 Miles"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="key"
    >
      <div
        className="radio-button"
      >
        <div
          className="hide"
        />
      </div>
      <p
        className="key-text"
      >
        2000 Miles
      </p>
      <div
        className="optional-text"
      />
    </div>
    <p
      className="value"
    >
      $1633
    </p>
  </div>
  <div
    className="data-selector"
    key="200 Miles"
    onClick={[Function]}
    role="presentation"
  >
    <div
      className="key"
    >
      <div
        className="radio-button"
      >
        <div
          className="hide"
        />
      </div>
      <p
        className="key-text"
      >
        200 Miles
      </p>
      <div
        className="optional-text"
      />
    </div>
    <p
      className="value"
    >
      $633
    </p>
  </div>
</React.Fragment>
`;
